### PR TITLE
Bratseth/input fallback

### PR DESCRIPTION
**Pull request by bratseth**

### CODE REVIEW

File: indexinglanguage/src/main/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceExpression.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
+package com.yahoo.vespa.indexinglanguage.expressions;
```
Comment (geirst) [inline indexinglanguage/src/main/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceExpression.java:1]: Please add copyright.

File: indexinglanguage/src/main/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceExpression.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    private static DataType resolveInputType(Collection<? extends Expression> list) {
+        DataType previousInput = null;
+        DataType previousOutput = null;
+        for (Expression choice : list) {
+            DataType thisInput = choice.requiredInputType();
+            if (previousInput == null)
+                previousInput = thisInput;
+            else if (thisInput != null && !previousInput.isAssignableFrom(thisInput))
+                throw new VerificationException(ScriptExpression.class, "Choice expressions require conflicting input types, " +
```
Comment (geirst) [inline indexinglanguage/src/main/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceExpression.java:59]: Should the exception message be "require non-conflicting input types" instead?

File: indexinglanguage/src/test/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceTestCase.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package com.yahoo.vespa.indexinglanguage.expressions;
```
Comment (geirst) [inline indexinglanguage/src/test/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceTestCase.java:1]: Please add copyright.

File: indexinglanguage/src/main/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceExpression.java (hunk 2 above)
Comment (geirst) [inline indexinglanguage/src/main/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceExpression.java:59]: Consider adding unit test that triggers this exception.

File: indexinglanguage/src/main/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceExpression.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                                                                        previousInput.getName() + " vs " + thisInput.getName() + ".");
+
+            DataType thisOutput = choice.createdOutputType();
+            if (previousOutput == null)
+                previousOutput = thisOutput;
+            else if (thisOutput != null && !previousOutput.isAssignableFrom(thisOutput))
+                throw new VerificationException(ScriptExpression.class, "Choice expressions produce conflicting output types, " +
```
Comment (geirst) [inline indexinglanguage/src/main/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceExpression.java:66]: Consider adding unit test that triggers this exception.

File: docprocs/src/test/java/com/yahoo/docprocs/indexing/IndexingProcessorTestCase.java [hunk 5]
```diff
@@ -124,33 +82,42 @@ public void testFieldDependingOnTwoInputs() {
...
 
-    @Test
-    public void requireThatEmptyDocumentUpdateOutputDoesNotThrow() {
-        DocumentType inputType = indexer.getDocumentTypeManager().getDocumentType("music");
-        DocumentUpdate input = new DocumentUpdate(inputType, "id:ns:music::");
-        Processing proc = new Processing();
-        proc.getDocumentOperations().add(input);
-        indexer.process(proc);
-        assertEquals(0, proc.getDocumentOperations().size());
+        {   // Neither title nor artist is set: Should not update embeddings even though it has fallbacks for all
```
Comment (geirst) [inline docprocs/src/test/java/com/yahoo/docprocs/indexing/IndexingProcessorTestCase.java:103]: We should have a similar set of tests for put operations. For this case I think combinedWithFallback should be set to an empty string if it is a put operation.

File: indexinglanguage/src/test/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceTestCase.java (hunk 3 above)
Comment (bratseth) [reply indexinglanguage/src/test/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceTestCase.java:2]: done

File: indexinglanguage/src/main/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceExpression.java (hunk 2 above)
Comment (bratseth) [reply indexinglanguage/src/main/java/com/yahoo/vespa/indexinglanguage/expressions/ChoiceExpression.java]: done

File: docprocs/src/test/java/com/yahoo/docprocs/indexing/IndexingProcessorTestCase.java (hunk 5 above)
Comment (bratseth) [reply docprocs/src/test/java/com/yahoo/docprocs/indexing/IndexingProcessorTestCase.java:160]: Yes, done.

**Review by geirst — COMMENTED** [review]
👍

**Pull request merged**